src/native/libs/System.Native/pal_interfaceaddresses.c

```diff
@@ -95,14 +100,75 @@ static inline uint8_t mask2prefix(uint8_t* mask, int length)
 
     return len;
 }
+
+#if !HAVE_IFADDRS && TARGET_ANDROID
+// This structure is exactly the same as struct ifaddrs defined in ifaddrs.h but since the header
+// might not be available (e.g., in bionics used in Android before API 24) we need to mirror it here
+// so that we can dynamically load the getifaddrs function and use it.
+struct ifaddrs
+{
+	struct ifaddrs *ifa_next;
+	char *ifa_name;
+	unsigned int ifa_flags;
+	struct sockaddr *ifa_addr;
+	struct sockaddr *ifa_netmask;
+	union
+	{
+		struct sockaddr *ifu_broadaddr;
+		struct sockaddr *ifu_dstaddr;
+	} ifa_ifu;
+	void *ifa_data;
+};
+#endif
+
+#if !HAVE_GETIFADDRS && TARGET_ANDROID
+// Try to load the getifaddrs and freeifaddrs functions manually.
+// This workaround is necessary on Android prior to API 24 and it can be removed once
+// we drop support for earlier Android versions.
+static int (*getifaddrs)(struct ifaddrs**) = NULL;
+static void (*freeifaddrs)(struct ifaddrs*) = NULL;
+
+static bool ensure_getifaddrs_is_loaded()
+{
+    static bool loading_already_attempted = false;
+    static pthread_mutex_t lock = PTHREAD_MUTEX_INITIALIZER;
+
+    if (pthread_mutex_lock(&lock) == 0)
+    {
+        if (!loading_already_attempted)
+        {
+            void *libc = dlopen("libc.so", RTLD_NOW);
```

Comment:
Would have to be tested, because bionic is known to have had issues with dynamic libraries and symbols. TBH, I wouldn't risk using weak symbols on Android (even though it's a good idea in general)

Reply:
I tried using weak symbols and it works just fine on my phone + all the Android tests passed and printed correct information. I'm not sure if it'll work on all Android versions though (and how to verify it).

Reply:
There's no way you can verify it 100%, alas. OEMs modify their Androids in a myriad ways and some of them are known to break more stuff than others, so it's better to avoid using things that may break. However, if you could create API 21 through 31 emulators and run the tests on these. That would at least prove it works on base Android. I would still avoid using the weak symbols, just to err on the side of caution.